disnake/ext/commands/interaction_bot_base.py

```diff
@@ -231,32 +254,123 @@ def command_sync_flags(self) -> CommandSyncFlags:
+        # no docstring because it was an attribute and now it's deprecated
+        return {
+            cmd.name: cmd
+            for cmd in self.all_app_commands.values()
+            if isinstance(cmd, InvokableSlashCommand)
+        }
+
+    @property
+    @deprecated("user_commands")
+    def all_user_commands(self) -> Dict[str, InvokableUserCommand]:
+        # no docstring because it was an attribute and now it's deprecated
+        return {
+            cmd.name: cmd
+            for cmd in self.all_app_commands.values()
+            if isinstance(cmd, InvokableUserCommand)
+        }
+
+    @property
+    @deprecated("message_commands")
+    def all_message_commands(self) -> Dict[str, InvokableMessageCommand]:
+        # no docstring because it was an attribute and now it's deprecated
+        return {
+            cmd.name: cmd
+            for cmd in self.all_app_commands.values()
+            if isinstance(cmd, InvokableMessageCommand)
+        }
+
+    def add_app_command(self, app_command: InvokableApplicationCommand) -> None:
+        """Adds an :class:`InvokableApplicationCommand` into the internal list of app commands.
+
+        This is usually not called, instead shortcut decorators are used, such as
+        :meth:`.slash_command`, :meth:`.user_command` or :meth:`.message_command`.
+
+        The app command is registered to guilds specified in the ``guild_ids`` attribute.
+        If this attribute is ``None`` then the command is registered globally, unless
+        parameter ``test_guilds`` is specified in the bot constructor, in which case
+        this command is registered to those guilds.
+
+        .. versionadded:: 2.10
+
+        Parameters
+        ----------
+        app_command: :class:`InvokableApplicationCommand`
+            The app command to add.
+
+        Raises
+        ------
+        ApplicationCommandRegistrationError
+            The app command is already registered.
+        TypeError
+            The app command passed is not an instance of :class:`InvokableApplicationCommand`.
+        """
+        if not isinstance(self, disnake.Client):
+            raise NotImplementedError("This method is only usable in disnake.Client subclasses")
+
+        if not isinstance(app_command, InvokableApplicationCommand):
+            raise TypeError(
+                "The app_command passed must be an instance of InvokableApplicationCommand"
+            )
+        if isinstance(app_command, (SubCommand, SubCommandGroup)):
+            raise TypeError(
+                "The app_command passed must be a top level command, "
+                "not an instance of SubCommand or SubCommandGroup"
+            )
+
+        if app_command.guild_ids is None:
+            # if test_guilds are specified then we add the same command for each test guild
+            guild_ids = (None,) if self._test_guilds is None else self._test_guilds
+        else:
+            guild_ids = app_command.guild_ids
```

Comment:
This snippet seems like a good candidate to be moved into its own (module-level?) function, it's used at least twice

Reply:
This snippet is no longer used because it caused a breaking change.